File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    answer = 0
    countList = [set() for _ in range(9)]
    countList[1].add(N)
    
    for i in range(2, 9):
        countset = set()
        for j in range(1,i):
            pre = countList[j]
            post = countList[i-j]
            
            for prenum in pre:
                for postnum in post:
                    countset.add(prenum+postnum)
                    countset.add(prenum-postnum)
                    countset.add(postnum-prenum)
                    countset.add(prenum*postnum)
                    if postnum != 0:
                        countset.add(prenum//postnum)
                    if prenum != 0: 
                        countset.add(postnum//prenum)
        countset.add(int(str(N)*i))
        countList[i] = countset
    
    for i in range(1,9):
        if number in countList[i]:
            return i
    
    return -1

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.cpp
#include <string>
#include <vector>
#include <set>
using namespace std;
int solution(int N, int number) {
    int answer = 0;
    int repeat = 0;
    vector<set<int>> dp(9);
    
    for(int i=1; i<=8; i++){
        repeat = repeat*10 +N;
        dp[i].insert(repeat);
    }
    for(int i=1; i<9; i++){
        for(int j=1; j<i; j++){
            for(auto a1: dp[j]){
                for(auto b1: dp[i-j]){
                    dp[i].insert(a1+b1);
                    if(a1>b1)   dp[i].insert(a1-b1);
                    else    dp[i].insert(-a1+b1);
                    dp[i].insert(a1*b1);
                    if(b1!=0) dp[i].insert(a1/b1);
                }
            }
        }
        if(dp[i].find(number)!=dp[i].end()) return i;
    }
    
    return -1;
}